tools/io-trace-explorer/io_trace.cc

```diff
@@ -0,0 +1,415 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <vector>
+
+#include "io_trace.h"
+
+/*
+ * io_trace::io_trace --
+ *     Initialize a new io_trace object.
+ */
+io_trace::io_trace(io_trace_collection &parent, const char *name) : _name(name), _parent(parent) {}
+
+/*
+ * io_trace::~io_trace --
+ *     Destroy a io_trace object.
+ */
+io_trace::~io_trace() {}
+
+/*
+ * io_trace_collection::io_trace_collection --
+ *     Initialize a new trace collection.
+ */
+io_trace_collection::io_trace_collection() {}
+
+/*
+ * io_trace_collection::~io_trace_collection --
+ *     Destroy a collection of traces.
+ */
+io_trace_collection::~io_trace_collection()
+{
+    for (auto &e : _traces) {
+        delete e.second;
+    }
+}
+
+/*
+ * io_trace_collection::add --
+ *     Add a data point.
+ */
+void
+io_trace_collection::add_data_point(
+  const std::string &device_or_file, const io_trace_kind kind, const io_trace_operation &item)
+{
+    std::string name = device_or_file;
+
+    /* Postprocess the data point. */
+
+    if (kind == io_trace_kind::WIRED_TIGER) {
+        const char *s = name.c_str();
+        if (strstr(s, "Log.00000") || strstr(s, "log.00000")) {
+            name = "WiredTiger Logs (combined)";
+        } else if (strlen(s) > 13 && strcmp(s + (strlen(s) - 13), "WiredTiger.wt") == 0) {
+            name = std::string("WiredTiger Main File: ") + s;
+        } else if (strlen(s) > 15 && strcmp(s + (strlen(s) - 15), "WiredTigerHS.wt") == 0) {
+            name = std::string("WiredTiger Main File: ") + s;
+        } else if (strlen(s) > 3 && strcmp(s + (strlen(s) - 3), ".wt") == 0) {
+            name = std::string("WiredTiger File: ") + s;
+        } else {
+            name = "WiredTiger Other";
+        }
+    }
+
+    if (kind == io_trace_kind::DEVICE) {
+        name = "Raw Device: " + name;
+    }
+
+    /* Find the correct trace. */
+
+    auto ti = _traces.find(name);
+    io_trace *t;
+    if (ti == _traces.end()) {
+        t = new io_trace(*this, name.c_str());
+        _traces[name] = t;
+    } else {
+        t = ti->second;
+    }
+
+    /* Insert the item into the correct place to keep the list sorted. */
+
+    if (t->_operations.empty())
+        t->_operations.push_back(item);
+
+    else {
+        const io_trace_operation &last = t->_operations.back();
+        if (last.timestamp <= item.timestamp)
+            t->_operations.push_back(item);
+        else
+            t->_operations.insert(
+              std::upper_bound(t->_operations.begin(), t->_operations.end(), item), item);
+    }
+}
+
+/*
+ * io_trace_collection::load_from_file --
+ *     Load a collection from file.
+ */
+void
+io_trace_collection::load_from_file(const char *file)
```

Comment:
It looks to me like all callers to this have a std::string, any reason `char *` is preferable here?